// jakes-gordon-growing-packer/build/sort-method.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
var sort = {
	random: function random(a, b) {
		return Math.random() - 0.5;
	},
	w: function w(a, b) {
		return b.w - a.w;
	},
	h: function h(a, b) {
		return b.h - a.h;
	},
	a: function a(_a, b) {
		return b.area - _a.area;
	},
	max: function max(a, b) {
		return Math.max(b.w, b.h) - Math.max(a.w, a.h);
	},
	min: function min(a, b) {
		return Math.min(b.w, b.h) - Math.min(a.w, a.h);
	},
	height: function height(a, b) {
		return msort(a, b, ['h', 'w']);
	},
	width: function width(a, b) {
		return msort(a, b, ['w', 'h']);
	},
	area: function area(a, b) {
		return msort(a, b, ['a', 'h', 'w']);
	},
	maxside: function maxside(a, b) {
		return msort(a, b, ['max', 'min', 'h', 'w']);
	}
};

function msort(a, b, criteria) {
	/* sort by multiple criteria */
	var diff = void 0,
	    n = void 0;
	for (n = 0; n < criteria.length; n++) {
		diff = sort[criteria[n]](a, b);

		if (diff != 0) {
			return diff;
		}
	}
	return 0;
}

exports.default = sort;
